content/news.md

```diff
@@ -4,6 +4,50 @@ title = "News"
 
 # News
 
+## 20180615 | PinTracker Revamp | @lanzafame
+
+In this post, I am going to cover how the `ipfs-cluster`'s PinTracker component used to work, what some of the issues with that implementation, how we fixed them, and where to next.
+
+#### How the PinTracker worked
+First the purpose of the pintracker, the pintracker serves the role of ferrying the appropriate state from ipfs-cluster's shared state to a node's ipfs daemon state.
+
+How this occurs is as follows:
+- ipfs-cluster receives a request to Pin a particular Cid
+- this request is routed to the consensus component, where it is stored in the distributed log and then an RPC call to Track the Cid is made to the pintracker
+- the pintracker then creates and stores the PinInfo in an internal map, before making a Pin request to the IPFS node via an RPC call to the IPFSConnector component
+- the IPFSConnector component is what finally requests the ipfs daemon to pin the Cid
+
+#### Issues that we faced
+The issues are separated into those which were due to how we initially implemented the MapPinTracker and then those that were/are inherent in any implementation of the pintracker that uses a map internally to store the status of the pins.
+
+Issues with the implementation:
+
+- the local state would get stale and required repeated syncing
+- race conditions
+- inability to cancel ongoing operations
+  - resulted in unnecessary requests to 'undo' requests that couldn't be cancelled
+- no way of knowing how many requests were currently in flight/queued for a single Cid
+
+Issues with a MapPinTracker:
+
+- local state (the map inside the pintracker) is potential to large to keep for large clusters
+- unnecessary duplication of PinInfo in the pintracker component
+
+#### How we fixed it
+To tackle the issues with current implementation of the MapPinTracker, we did the following things. 
+
+We moved to a model where we track in-flight operations, so instead of a map that stored the status of a Cid, i.e. `map[Cid]PinInfo`, we now store an operation for a Cid, i.e. `map[Cid]Operation`. 
+
+Now, an Operation contains not only the type of operation that is being performed for a Cid, Pin or Unpin, and the phase of the operation, Queued, In Progress, Done, or Error, but a [`context.Context`](https://golang.org/pkg/context).
+
+Having a context available in every Operation gives us the ability to cancel an operation at any point but also upon receiving opposing operations for the same Cid we can cancel the in-flight operation automatically, maybe even before that operation had started to be processed depending on the timing. 
```

Comment:
Also need to mention that Recover is now async, that there is no real need to sync the shared state anymore as regularly as before because it's way more difficult that they are out of sync and that we can control pinning and unpinning timeouts in the ipfs connector and that unlike before, they will cancel the operation.